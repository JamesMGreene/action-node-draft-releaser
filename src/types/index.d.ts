export interface ICommit {
  sha: string;
  url: string;
}

export interface ITag {
  commit: ICommit;
  name: string;
  tarball_url: string;
  zipball_url: string;
}

export interface IUser {
  id: number;
  login: string;
  // etc.
}

export interface IRelease {
  assets: any[];
  assets_url: string;
  author: IUser;
  body: string;
  created_at: string;
  draft: boolean;
  html_url: string;
  id: number;
  name: string;
  prerelease: boolean;
  published_at: string | null;
  tag_name: string;
  tarball_url: string | null;
  target_commitish: string;
  upload_url: string;
  url: string;
  zipball_url: string | null;
}
