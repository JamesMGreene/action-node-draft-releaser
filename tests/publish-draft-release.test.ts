import nock from 'nock';
import publishDraftRelease from '../src/publish-draft-release';
import { IRelease } from '../src/types/index';

const owner = 'JamesMGreene';
const repo = 'test-repo';

let releaseIdCounter = 0;
function generateDraftRelease (version: string, name: string = version): IRelease {
  const releaseId = ++releaseIdCounter;
  return {
    assets: [],
    assets_url: `https://api.github.com/repos/${owner}/${repo}/releases/${releaseId}/assets`,
    author: {
      id: 417751,
      login: owner
    },
    body: 'Blah',
    created_at: '2019-08-18T03:34:57Z',
    draft: true,
    html_url: `https://github.com/${owner}/${repo}/releases/tag/untagged-29db7ff348b9db3f003e`,
    id: releaseId,
    name,
    prerelease: false,
    published_at: null,
    tag_name: version,
    tarball_url: null,
    target_commitish: 'master',
    upload_url: `https://uploads.github.com/repos/${owner}/${repo}/releases/${releaseId}/assets{?name,label}`,
    url: `https://api.github.com/repos/${owner}/${repo}/releases/${releaseId}`,
    zipball_url: null
  };
}

describe('publishDraftRelease', () => {
  const someVersion = '2.0.0';
  const prereleaseVersion = '2.0.0-beta.1';

  const someDraftRelease = generateDraftRelease(someVersion);
  const someDraftReleaseWithLesserVersionName = generateDraftRelease(someVersion, prereleaseVersion);
  const someDraftReleaseWithGreaterVersionName = generateDraftRelease(someVersion, '3.0.0');
  const someDraftReleaseWithNonVersionName = generateDraftRelease(someVersion, 'The Big One!');

  const draftPrerelease = generateDraftRelease(prereleaseVersion);

  beforeEach(() => {
    process.env.GITHUB_TOKEN = 'testToken123';

    // To enable: `import { context } from '@actions/github'`
    process.env.GITHUB_REPOSITORY = `${owner}/${repo}`;

    // For creating the Release and Tag from the pushed commit SHA
    process.env.GITHUB_SHA = 'deadbeefcafe1c6e148a9be56893bed50abadbed';
  });

  afterEach(() => {
    delete process.env.GITHUB_SHA;
    delete process.env.GITHUB_REPOSITORY;
    delete process.env.GITHUB_TOKEN;
  });

  describe('draft published as normal release', () => {
    it('should update as expected', async () => {
      nock('https://api.github.com')
        .patch(`/repos/${owner}/${repo}/releases/${someDraftRelease.id}`)
          .reply(200, (uri, requestBody) => ({
            ...someDraftRelease,
            ...requestBody as object,
            published_at: '2019-08-19T03:34:57Z'
          }));

      const release = await publishDraftRelease(someDraftRelease, someVersion, true);
      expect(release).toMatchSnapshot();
      expect(release.name).toBe(someVersion);
      expect(release.name).toBe(someDraftRelease.name);
    });

    describe('with allowNameUpdate:true', () => {
      it('should update the release name if it is a lesser version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithLesserVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithLesserVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithLesserVersionName, someVersion, true);
        expect(release).toMatchSnapshot();
        expect(release.name).toBe(someVersion);
        expect(release.name).not.toBe(someDraftReleaseWithLesserVersionName.name);
      });

      it('should update the release name if it is a greater version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithGreaterVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithGreaterVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithGreaterVersionName, someVersion, true);
        expect(release).toMatchSnapshot();
        expect(release.name).toBe(someVersion);
        expect(release.name).not.toBe(someDraftReleaseWithGreaterVersionName.name);
      });

      it('should not update the release name if it is not a version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithNonVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithNonVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithNonVersionName, someVersion, true);
        expect(release).toMatchSnapshot();
        expect(release.name).not.toBe(someVersion);
        expect(release.name).toBe(someDraftReleaseWithNonVersionName.name);
      });
    });

    describe('with allowNameUpdate:false', () => {
      it('should not update the release name if it is a lesser version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithLesserVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithLesserVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithLesserVersionName, someVersion, false);
        expect(release).toMatchSnapshot();
        expect(release.name).not.toBe(someVersion);
        expect(release.name).toBe(someDraftReleaseWithLesserVersionName.name);
      });

      it('should not update the release name if it is a greater version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithGreaterVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithGreaterVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithGreaterVersionName, someVersion, false);
        expect(release).toMatchSnapshot();
        expect(release.name).not.toBe(someVersion);
        expect(release.name).toBe(someDraftReleaseWithGreaterVersionName.name);
      });

      it('should not update the release name if it is not a version', async () => {
        nock('https://api.github.com')
          .patch(`/repos/${owner}/${repo}/releases/${someDraftReleaseWithNonVersionName.id}`)
            .reply(200, (uri, requestBody) => ({
              ...someDraftReleaseWithNonVersionName,
              ...requestBody as object,
              published_at: '2019-08-19T03:34:57Z'
            }));

        const release = await publishDraftRelease(someDraftReleaseWithNonVersionName, someVersion, false);
        expect(release).toMatchSnapshot();
        expect(release.name).not.toBe(someVersion);
        expect(release.name).toBe(someDraftReleaseWithNonVersionName.name);
      });
    });
  });

  describe('draft published as prerelease', () => {
    it('should update as expected', async () => {
      nock('https://api.github.com')
        .patch(`/repos/${owner}/${repo}/releases/${draftPrerelease.id}`)
          .reply(200, (uri, requestBody) => ({
            ...draftPrerelease,
            ...requestBody as object,
            published_at: '2019-08-19T03:34:57Z'
          }));

      const prerelease = await publishDraftRelease(draftPrerelease, prereleaseVersion, true);
      expect(prerelease).toMatchSnapshot();
      expect(prerelease.name).toBe(draftPrerelease.name);
    });
  });

  describe('not found error', () => {
    it('should throw', async () => {
      nock('https://api.github.com')
        .patch(`/repos/${owner}/${repo}/releases/${someDraftRelease.id}`)
          .reply(404);

      await expect(publishDraftRelease(someDraftRelease, someVersion, true)).rejects.toThrowError();
    });
  });

  describe('other error', () => {
    it('should throw', async () => {
      nock('https://api.github.com')
        .patch(`/repos/${owner}/${repo}/releases/${someDraftRelease.id}`)
          .reply(401);

      await expect(publishDraftRelease(someDraftRelease, someVersion, true)).rejects.toThrowError();
    });
  });
});
