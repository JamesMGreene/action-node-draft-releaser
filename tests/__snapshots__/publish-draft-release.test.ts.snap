// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publishDraftRelease draft published as normal release should update as expected 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/1/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 1,
  "name": "2.0.0",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/1/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/1",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:false should not update the release name if it is a greater version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/3/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 3,
  "name": "3.0.0",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/3/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/3",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:false should not update the release name if it is a lesser version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/2/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 2,
  "name": "2.0.0-beta.1",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/2/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/2",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:false should not update the release name if it is not a version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/4/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 4,
  "name": "The Big One!",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/4/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/4",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:true should not update the release name if it is not a version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/4/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 4,
  "name": "The Big One!",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/4/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/4",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:true should update the release name if it is a greater version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/3/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 3,
  "name": "2.0.0",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/3/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/3",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as normal release with allowNameUpdate:true should update the release name if it is a lesser version 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/2/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 2,
  "name": "2.0.0",
  "prerelease": false,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/2/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/2",
  "zipball_url": null,
}
`;

exports[`publishDraftRelease draft published as prerelease should update as expected 1`] = `
Object {
  "assets": Array [],
  "assets_url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/5/assets",
  "author": Object {
    "id": 417751,
    "login": "JamesMGreene",
  },
  "body": "Blah",
  "created_at": "2019-08-18T03:34:57Z",
  "draft": false,
  "html_url": "https://github.com/JamesMGreene/test-repo/releases/tag/untagged-29db7ff348b9db3f003e",
  "id": 5,
  "name": "2.0.0-beta.1",
  "prerelease": true,
  "published_at": "2019-08-19T03:34:57Z",
  "tag_name": "2.0.0-beta.1",
  "tarball_url": null,
  "target_commitish": "deadbeefcafe1c6e148a9be56893bed50abadbed",
  "upload_url": "https://uploads.github.com/repos/JamesMGreene/test-repo/releases/5/assets{?name,label}",
  "url": "https://api.github.com/repos/JamesMGreene/test-repo/releases/5",
  "zipball_url": null,
}
`;
